sdk/helper/keysutil/policy.go

```diff
@@ -2392,3 +2395,201 @@ func wrapTargetPKCS8ForImport(wrappingKey *rsa.PublicKey, preppedTargetKey []byt
+
+func (p *Policy) CreateCSR(keyVersion int, csrTemplate *x509.CertificateRequest) ([]byte, error) {
+	if !p.Type.SigningSupported() {
+		return nil, errutil.UserError{Err: fmt.Sprintf("key type '%s' does not support signing", p.Type)}
+	}
+
+	keyEntry, err := p.safeGetKeyEntry(keyVersion)
+	if err != nil {
+		return nil, err
+	}
+
+	if keyEntry.IsPrivateKeyMissing() {
+		return nil, errutil.UserError{Err: "private key not imported for key version selected"}
+	}
+
+	csrTemplate.Signature = nil
+	csrTemplate.SignatureAlgorithm = x509.UnknownSignatureAlgorithm
+
+	key, err := p.getPrivateKey(&keyEntry)
+	if err != nil {
+		return nil, err
+	}
+
+	csrBytes, err := x509.CreateCertificateRequest(rand.Reader, csrTemplate, key)
+	if err != nil {
+		return nil, fmt.Errorf("could not create the CSR: %w", err)
+	}
+
+	pemCSR := pem.EncodeToMemory(&pem.Block{
+		Type:  "CERTIFICATE REQUEST",
+		Bytes: csrBytes,
+	})
+
+	return pemCSR, nil
+}
+
+func (p *Policy) getPrivateKey(keyEntry *KeyEntry) (crypto.Signer, error) {
+	switch p.Type {
+	case KeyType_ECDSA_P256, KeyType_ECDSA_P384, KeyType_ECDSA_P521:
+		curve, err := p.getECDSAKeyCurve()
+		if err != nil {
+			return nil, err
+		}
+		key := &ecdsa.PrivateKey{
+			PublicKey: ecdsa.PublicKey{
+				Curve: curve,
+				X:     keyEntry.EC_X,
+				Y:     keyEntry.EC_Y,
+			},
+			D: keyEntry.EC_D,
+		}
+		return key, nil
+	case KeyType_ED25519:
+		if p.Derived {
+			return nil, errutil.UserError{Err: "operation not supported on keys with derivation enabled"}
+		}
+		return ed25519.PrivateKey(keyEntry.Key), nil
+	case KeyType_RSA2048, KeyType_RSA3072, KeyType_RSA4096:
+		return keyEntry.RSAKey, nil
+	default:
+		return nil, errutil.InternalError{Err: fmt.Sprintf("selected key type '%s' does not support signing", p.Type.String())}
+	}
+}
+
+func (p *Policy) getECDSAKeyCurve() (elliptic.Curve, error) {
+	switch p.Type {
+	case KeyType_ECDSA_P256:
+		return elliptic.P256(), nil
+	case KeyType_ECDSA_P384:
+		return elliptic.P384(), nil
+	case KeyType_ECDSA_P521:
+		return elliptic.P521(), nil
+	default:
+		return nil, errutil.InternalError{Err: fmt.Sprintf("key type '%s' does not have a curve parameter", p.Type)}
+	}
+}
+
+func (p *Policy) PersistCertificateChain(ctx context.Context, storage logical.Storage, keyVersion int, certificateChain []*x509.Certificate) error {
+	// validate that the certificate chain has at least one certifiate, the leaf certificate
+	if len(certificateChain) == 0 {
+		return errutil.UserError{Err: "expected at least one certificate in the certificate chain"}
+	}
+
+	// validate that the first element in the certificate chain is a leaf certificate
+	if certificateChain[0].IsCA && certificateChain[0].BasicConstraintsValid {
+		return errutil.UserError{Err: "certificate in the first element is not a valid leaf certificate"}
+	}
+
+	// validate that the last element in the certificate chain is a CA certificate
+	if len(certificateChain) > 1 && !certificateChain[len(certificateChain)-1].IsCA {
```

Comment:
```suggestion
	if len(certificateChain) > 1 && (!certificateChain[len(certificateChain)-1].BasicConstraintsValid || certificateChain[len(certificateChain)-1].IsCA) {
```

Reply:
Even if checking for `false`? I assumed that a check for `true` required checking the basic constraints, if not, it wasn't necessary.

Reply:
In that case we might want the checks like the following, correct? If the BC value isn't valid, `true`, we can't trust the `IsCA` value so cannot properly evaluate the certificate "type".
```go
	// validate that the first element in the certificate chain is a leaf certificate
	if !certificateChain[0].BasicConstraintsValid || certificateChain[0].IsCA {
		return errutil.UserError{Err: "certificate in the first element is not a valid leaf certificate"}
	}

	// validate that the last element in the certificate chain is a CA certificate
	if len(certificateChain) > 1 && (!certificateChain[len(certificateChain)-1].BasicConstraintsValid || !certificateChain[len(certificateChain)-1].IsCA) {
		return errutil.UserError{Err: "certificate in the last element is not a valid CA certificate"}
	}

	// validate that the certificate chain contains only one leaf certificate
	for _, certificate := range certificateChain[1 : len(certificateChain)-1] {
		if !certificate.BasicConstraintsValid || !certificate.IsCA {
			return errutil.UserError{Err: "provided certificate chain contains more than one leaf certificate"}
		}
	}

```

Reply:
@Gabrielopesantos No, basic constraint not being valid implies that [it is _not_ a CA](https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.9).

So the first check will also match on valid leaf certificates that don't have a basic constraint extension.

The second check is correct, because we want to fail all leaf certs: those without a basic constraint extension, or those with a basic constraint extension and IsCA asserted to true. The for loop's statement is correct for the same reason. :-)

Technically, the for loop duplicates the same as the second check, so we could probably just broaden the range to `certificateChain[1:len(certificateChain)]` perhaps?